File: scripts/next_major_doctor.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = ["pyyaml"]
# ///
"""Head-of-output next-major doctor section (authority: references/cli/update-channels.yaml)."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from yaml_mapping import load_yaml_mapping_file

UPDATE_CHANNELS_AUTHORITY = Path("references/cli/update-channels.yaml")
NEXT_MAJOR_SECTION_HEADER = "Next major"
NEXT_MAJOR_LINE_CAP = 6
CHANNEL_NAMES = frozenset({"stable", "development"})

V1_NEXT_MAJOR_FALLBACK: dict[str, str] = {
    "concept": "forward_successor_line",
    "current_version": "1.x",
    "current_channel": "stable",
    "channel": "stable",
    "version": "2.x",
    "npm_only_advisory": "",
    "guide_url": "https://github.com/jgabor/agentera/blob/main/UPGRADE.md#recommended-upgrade-v1--v2-stable-channel",
    "preview_command": "npx -y agentera@latest upgrade --dry-run",
    "irreversible_advisory": (
        "Forward migration upgrades artifact format from Markdown to YAML; "
        "review the preview before applying."
    ),
}


def _authority_path(source_root: Path) -> Path:
    candidate = source_root / UPDATE_CHANNELS_AUTHORITY
    if candidate.exists():
        return candidate
    return Path(__file__).resolve().parents[1] / UPDATE_CHANNELS_AUTHORITY


def load_update_channels_authority(source_root: Path) -> dict[str, Any]:
    return load_yaml_mapping_file(_authority_path(source_root))


def _parse_next_major_block(raw: Any) -> dict[str, str] | None:
    if raw is None:
        return None
    if not isinstance(raw, dict):
        return None
    channel = str(raw.get("channel", "")).strip().lower()
    if channel not in CHANNEL_NAMES:
        return None
    version = str(raw.get("version", "")).strip()
    if not version:
        return None
    guide_url = str(raw.get("guide_url", "")).strip()
    preview_command = str(raw.get("preview_command", "")).strip()
    irreversible_advisory = str(raw.get("irreversible_advisory", "")).strip()
    if not guide_url or not preview_command or not irreversible_advisory:
        return None
    return {
        "concept": str(raw.get("concept", "forward_successor_line")).strip(),
        "channel": channel,
        "version": version,
        "npm_only_advisory": str(raw.get("npm_only_advisory", "")).strip(),
        "guide_url": guide_url,
        "preview_command": preview_command,
        "irreversible_advisory": irreversible_advisory,
    }


def load_channel_next_major(source_root: Path, channel: str) -> dict[str, str] | None:
    authority = load_update_channels_authority(source_root)
    channels = authority.get("channels")
    if not isinstance(channels, dict):
        return None
    entry = channels.get(channel)
    if not isinstance(entry, dict):
        return None
    return _parse_next_major_block(entry.get("next_major"))


def format_next_major_doctor_lines(
    *,
    current_version: str,
    current_channel: str,
    block: dict[str, str],
) -> list[str]:
    next_line = (
        f"Next: {block['version']} ({block['channel']} channel). {block['npm_only_advisory']}"
        if block.get("npm_only_advisory")
        else f"Next: {block['version']} ({block['channel']} channel)"
    )
    return [
        NEXT_MAJOR_SECTION_HEADER,
        f"Current: {current_version} ({current_channel} channel)",
        next_line,
        f"Guide: {block['guide_url']}",
        f"Preview: {block['preview_command']}",
        block["irreversible_advisory"],
    ]


def _running_distribution_major(running_version: str | None) -> int:
    if not running_version:
        return 2
    head = running_version.strip().split(".", 1)[0]
    if head.isdigit():
        return int(head)
    return 2


def resolve_next_major_doctor_lines(
    *,
    source_root: Path,
    channel: str = "stable",
    running_version: str | None = None,
    running_distribution_major: int | None = None,
) -> list[str] | None:
    current_channel = channel if channel in CHANNEL_NAMES else "stable"
    major = running_distribution_major if running_distribution_major is not None else _running_distribution_major(
        running_version
    )

    if major == 1:
        block = {**V1_NEXT_MAJOR_FALLBACK}
        return format_next_major_doctor_lines(
            current_version=block["current_version"],
            current_channel=block["current_channel"],
            block=block,
        )

    authority_block = load_channel_next_major(source_root, current_channel)
    if not authority_block:
        return None

    current_version = running_version or f"{major}.x"
    return format_next_major_doctor_lines(
        current_version=current_version,
        current_channel=current_channel,
        block=authority_block,
    )


def prepend_next_major_doctor_section(text: str, section_lines: list[str] | None) -> str:
    if not section_lines:
        return text
    return "\n".join(section_lines) + "\n\n" + text

File: packages/cli/src/upgrade/nextMajorDoctor.ts
import { cliDistributionMajor } from "./compatibility.js";
import {
  loadUpdateChannelsAuthority,
  resolveSelectedChannel,
  type UpdateChannelName,
} from "./channels.js";
import { resolveRunningVersion } from "./versionResolution.js";
import type { InstallClassification } from "./compatibility.js";

/**
 * Head-of-output "next major" doctor section.
 * Authority: references/cli/update-channels.yaml (channels.<channel>.next_major)
 */

export const NEXT_MAJOR_SECTION_HEADER = "Next major";
export const NEXT_MAJOR_LINE_CAP = 6;

export interface NextMajorBlock {
  concept: string;
  channel: UpdateChannelName;
  version: string;
  /** When false, doctor/prime omit successor-line upgrade prompts until republished. */
  announced: boolean;
  npmOnlyAdvisory: string;
  guideUrl: string;
  previewCommand: string;
  irreversibleAdvisory: string;
}

export interface NextMajorDoctorContext {
  sourceRoot: string;
  home: string;
  channel?: UpdateChannelName | null;
  runningVersion?: string | null;
  runningDistributionMajor?: number | null;
  install?: InstallClassification | null;
  env?: Record<string, string | undefined>;
}

/** v1 installs predate the channel authority; hardcoded stable-line successor mapping. */
export const V1_NEXT_MAJOR_FALLBACK: NextMajorBlock & { currentVersion: string; currentChannel: UpdateChannelName } = {
  concept: "forward_successor_line",
  currentVersion: "1.x",
  currentChannel: "stable",
  channel: "stable",
  version: "2.x",
  announced: true,
  npmOnlyAdvisory: "",
  guideUrl:
    "https://github.com/jgabor/agentera/blob/main/UPGRADE.md#recommended-upgrade-v1--v2-stable-channel",
  previewCommand: "npx -y agentera@latest upgrade --dry-run",
  irreversibleAdvisory:
    "Forward migration upgrades artifact format from Markdown to YAML; review the preview before applying.",
};

type Dict = Record<string, unknown>;

function parseNextMajorBlock(raw: unknown): NextMajorBlock | null {
  if (raw === null || raw === undefined) {
    return null;
  }
  if (typeof raw !== "object" || Array.isArray(raw)) {
    return null;
  }
  const block = raw as Dict;
  const channel = String(block.channel ?? "").trim().toLowerCase();
  if (channel !== "stable" && channel !== "development") {
    return null;
  }
  const version = String(block.version ?? "").trim();
  if (!version) {
    return null;
  }
  const guideUrl = String(block.guide_url ?? "").trim();
  const previewCommand = String(block.preview_command ?? "").trim();
  const irreversibleAdvisory = String(block.irreversible_advisory ?? "").trim();
  if (!guideUrl || !previewCommand || !irreversibleAdvisory) {
    return null;
  }
  return {
    concept: String(block.concept ?? "forward_successor_line").trim(),
    channel: channel as UpdateChannelName,
    version,
    announced: block.announced !== false,
    npmOnlyAdvisory: String(block.npm_only_advisory ?? "").trim(),
    guideUrl,
    previewCommand,
    irreversibleAdvisory,
  };
}

let testSuccessorAnnouncedOverride: boolean | null = null;

/** Reset injectable successor announcement gate (tests). */
export function setSuccessorAnnouncedOverrideForTests(value: boolean | null): void {
  testSuccessorAnnouncedOverride = value;
}

/**
 * True when the resolved channel's successor metadata exists and is announced
 * for doctor/prime. Defaults to the stable channel so legacy callers that
 * omit the argument behave identically to pre-fix behavior.
 */
export function isStableSuccessorAnnounced(
  sourceRoot: string,
  channel: UpdateChannelName = "stable",
): boolean {
  if (testSuccessorAnnouncedOverride !== null) {
    return testSuccessorAnnouncedOverride;
  }
  const block = loadChannelNextMajor(sourceRoot, channel);
  return block?.announced === true;
}

export function loadChannelNextMajor(sourceRoot: string, channel: UpdateChannelName): NextMajorBlock | null {
  const authority = loadUpdateChannelsAuthority(sourceRoot);
  const channels = authority.channels as Dict | undefined;
  const entry = channels?.[channel];
  if (!entry || typeof entry !== "object" || Array.isArray(entry)) {
    return null;
  }
  return parseNextMajorBlock((entry as Dict).next_major);
}

export function formatNextMajorDoctorLines(args: {
  currentVersion: string;
  currentChannel: UpdateChannelName;
  block: NextMajorBlock;
}): string[] {
  const nextLine = args.block.npmOnlyAdvisory
    ? `Next: ${args.block.version} (${args.block.channel} channel). ${args.block.npmOnlyAdvisory}`
    : `Next: ${args.block.version} (${args.block.channel} channel)`;
  return [
    NEXT_MAJOR_SECTION_HEADER,
    `Current: ${args.currentVersion} (${args.currentChannel} channel)`,
    nextLine,
    `Guide: ${args.block.guideUrl}`,
    `Preview: ${args.block.previewCommand}`,
    args.block.irreversibleAdvisory,
  ];
}

function installTrackSkipsSuccessorBlock(
  install: InstallClassification | null | undefined,
  sourceRoot: string,
): boolean {
  if (install?.kind === "v3_self_contained_npm") {
    return true;
  }
  if (install?.kind === "source_checkout" && cliDistributionMajor(sourceRoot) >= 3) {
    return true;
  }
  return false;
}

export function resolveNextMajorDoctorLines(ctx: NextMajorDoctorContext): string[] | null {
  const env = ctx.env ?? process.env;
  const home = ctx.home;
  const sourceRoot = ctx.sourceRoot;
  const install = ctx.install ?? null;

  if (installTrackSkipsSuccessorBlock(install, sourceRoot)) {
    return null;
  }

  const selected = resolveSelectedChannel({
    channel: ctx.channel ?? null,
    env,
    home,
    sourceRoot,
  });
  const currentChannel = selected.channel;
  const successorChannel: UpdateChannelName =
    install?.kind === "v2_managed_app_home" ? "stable" : currentChannel;
  const displayChannel: UpdateChannelName =
    install?.kind === "v2_managed_app_home" ? "stable" : currentChannel;

  const runningDistributionMajor =
    ctx.runningDistributionMajor ??
    (install?.kind === "v2_managed_app_home" ? 2 : cliDistributionMajor(sourceRoot));

  if (runningDistributionMajor === 1) {
    const block = V1_NEXT_MAJOR_FALLBACK;
    return formatNextMajorDoctorLines({
      currentVersion: block.currentVersion,
      currentChannel: block.currentChannel,
      block,
    });
  }

  const authorityBlock = loadChannelNextMajor(sourceRoot, successorChannel);
  if (!authorityBlock || !authorityBlock.announced) {
    return null;
  }

  const runningVersion =
    ctx.runningVersion ??
    (install
      ? resolveRunningVersion({
          appHome: install.appHome,
          sourceRoot,
          install,
        })
      : null) ??
    `${runningDistributionMajor}.x`;

  return formatNextMajorDoctorLines({
    currentVersion: runningVersion,
    currentChannel: displayChannel,
    block: authorityBlock,
  });
}

export function prependNextMajorDoctorSection(text: string, sectionLines: string[] | null): string {
  if (!sectionLines || sectionLines.length === 0) {
    return text;
  }
  return `${sectionLines.join("\n")}\n\n${text}`;
}
